### PR TITLE
Delegation

### LINKED ISSUE #2212: Support for delegation
We need support for delegation.

As we have two ledgers (staking and runtime) there are multiple options:
* Support delegation via the staking ledger in the consensus layer (works even if the runtimes are not working).
* Support delegation via the runtimes (similar to #2077).
* Support both.

Most likely at least part of this support is required in the consensus layer.

*More requirements and options TBD.*

### PR DESCRIPTION
**Pull request by kostko**
Fixes #2212 

TODO
* [x] Dump and restore delegation descriptors.
* [x] Dump and restore debonding delegation descriptors.
* [x] Make debonding period epoch-granular.

### CODE REVIEW

File: go/tendermint/apps/epochtime_mock/epochtime_mock.go [hunk 1]
```diff
@@ -149,7 +149,8 @@ func (app *epochTimeMockApplication) EndBlock(request types.RequestEndBlock) (ty
 	return types.ResponseEndBlock{}, nil
 }
 
-func (app *epochTimeMockApplication) FireTimer(ctx *abci.Context, timer *abci.Timer) {
+func (app *epochTimeMockApplication) FireTimer(ctx *abci.Context, timer *abci.Timer) error {
+	return nil
```
Comment (Yawning) [inline go/tendermint/apps/epochtime_mock/epochtime_mock.go]: Should this return an error?

File: go/tendermint/apps/staking/staking.go [hunk 2]
```diff
@@ -158,40 +158,42 @@ func (app *stakingApplication) InitChain(ctx *abci.Context, request types.Reques
...
 			)
 			return errors.New("staking/tendermint: invalid genesis escrow balance")
 		}
-
-		account := &ledgerEntry{
-			GeneralBalance:  v.GeneralBalance,
-			EscrowBalance:   v.EscrowBalance,
-			DebondStartTime: v.DebondStartTime,
-			Nonce:           v.Nonce,
+		if !v.Escrow.DebondingShares.IsZero() {
```
Comment (Yawning) [inline go/tendermint/apps/staking/staking.go]: Whowa, is this an ok assumption to make?  I don't think it is...

File: go/tendermint/apps/staking/staking.go [hunk 3]
```diff
@@ -323,15 +459,19 @@ func (app *stakingApplication) queryGenesis(s, r interface{}) ([]byte, error) {
...
 		acct := state.account(acctID)
-		ledger[acctID.ToMapKey()] = &staking.GenesisLedgerEntry{
-			GeneralBalance:  acct.GeneralBalance,
-			EscrowBalance:   acct.EscrowBalance,
-			DebondStartTime: acct.DebondStartTime,
-			Nonce:           acct.Nonce,
-		}
+		// All debonding stops on a dump/restore and needs to be restarted. This
+		// is also checked on chain init.
+		acct.Escrow.DebondingShares = *staking.NewQuantity()
```
Comment (Yawning) [inline go/tendermint/apps/staking/staking.go]: Ah I see.  I'm not sure people will be happy about this...

File: go/tendermint/apps/staking/staking.go [hunk 4]
```diff
@@ -542,63 +712,73 @@ func (app *stakingApplication) reclaimEscrow(ctx *abci.Context, state *MutableSt
...
-			"now", now,
-			"debond_at", debondAt,
+	// Fetch escrow account.
+	from := state.account(reclaim.Account)
+
+	// Fetch delegation.
+	delegation := state.delegation(id, reclaim.Account)
+
+	// Update delegated shares.
+	if err := delegation.Shares.Sub(&reclaim.Shares); err != nil {
```
Comment (Yawning) [inline go/tendermint/apps/staking/staking.go]: Hm, so "reclaim all my shares" requires querying the current share count?  We could use `SubUpTo` to make this easier, but not sure if being convenient is better than being explicit (No opinion either way).

File: go/tendermint/apps/staking/staking.go (hunk 3 above)
Comment (Yawning) [reply go/tendermint/apps/staking/staking.go]: Doesn't this eat tokens that are currently debonding?  The delegation's been updated, so the tokens held in debonding limbo need to either be put back, or this needs to be preserved no?

File: go/tendermint/apps/staking/staking.go (hunk 2 above)
Comment (kostko) [reply go/tendermint/apps/staking/staking.go]: Fixed so dump/restore now also dumps and restores unbonding delegations.

File: go/tendermint/apps/staking/staking.go (hunk 3 above)
Comment (kostko) [reply go/tendermint/apps/staking/staking.go]: Fixed so dump/restore now also dumps and restores unbonding delegations.

**Comment by kostko** [comment]
Changed debonding so that it is now epoch-based.

**Pull request merged**